sdk/nodejs/npm/manager.go

```diff
@@ -0,0 +1,107 @@
+package npm
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+)
+
+// A `PackageManager` is responsible for installing dependencies,
+// packaging Pulumi programs, and executing Node in the context of
+// installed packages. In practice, each implementation of this
+// interface represents one of the package managers in the Node ecosystem
+// e.g. Yarn, NPM, etc.
+// The language host will dynamically dispatch to an instance of PackageManager
+// in response to RPC requests.
+type PackageManager interface {
+	Install(ctx context.Context, dir string, production bool, stdout, stderr io.Writer) error
+	Pack(ctx context.Context, dir string, stderr io.Writer) ([]byte, error)
+}
+
+// Pack runs `npm pack` in the given directory, packaging the Node.js app located there into a
+// tarball and returning it as `[]byte`. `stdout` is ignored for the command, as it does not
+// generate useful data. If the `PULUMI_PREFER_YARN` environment variable is set, `yarn pack` is run
+// instead of `npm pack`.
+func Pack(ctx context.Context, dir string, stderr io.Writer) ([]byte, error) {
+	pkgManager, err := ResolvePackageManager(dir)
+	if err != nil {
+		return nil, err
+	}
+	return pkgManager.Pack(ctx, dir, stderr)
+}
+
+// Install runs `npm install` in the given directory, installing the dependencies for the Node.js
+// app located there. If the `PULUMI_PREFER_YARN` environment variable is set, `yarn install` is used
+// instead of `npm install`.
+func Install(ctx context.Context, dir string, production bool, stdout, stderr io.Writer) (string, error) {
+	bin := "npm"
+	pkgManager, err := ResolvePackageManager(dir)
+	if err != nil {
+		return bin, err
+	}
+
+	// TODO: It's not clear why Install returns the name of the package manager
+	// responsible for having performed the installation. But for now, we
+	// reflect on the type of the implementor. If we find we need to maintain
+	// this contract, we can add a `Name() string` function to the interface.
+	if _, ok := pkgManager.(*yarnClassic); ok {
+		bin = "yarn"
+	}
+
+	err = pkgManager.Install(ctx, dir, production, stdout, stderr)
+	if err != nil {
+		return bin, err
+	}
+
+	// Ensure the "node_modules" directory exists.
+	// NB: This is only approperate for certain package managers.
+	//     Yarn with Plug'n'Play enabled won't produce a node_modules directory,
+	//     either for Yarn Classic or Yarn Berry.
+	nodeModulesPath := filepath.Join(dir, "node_modules")
+	if _, err := os.Stat(nodeModulesPath); os.IsNotExist(err) {
```

Comment:
This is code from the previous incarnation ([link](https://github.com/pulumi/pulumi/pull/12983/files#diff-bf3f868f2d50fbb01087b440750d5fce4f5258db426ff8767bb7d54eb7a28398L99)). I guess it doesn't matter if you don't have permissions to stat the directory, because you might not be the user reading from `node_modules`?

Idk! What if I add a warning here for the other error cases? Not sure what's best.

Reply:
I added a warning, since it's definitely a bizarre scenario, but I don't think it's always a hard error.